### PR TITLE
New: array-bracket-spacing (fixes #2226)

### LINKED ISSUE #2226: Add spaces-in-square-brackets rule [$15]
As discussed in #1576, replacing space-in-brackets with individual `space-in-square-brackets` and `space-in-curly-braces` rules makes sense, adds flexibility and will reduce confusion going forward.

This issue is for tracking the addition of the `space-in-square-brackets` rule (which may need a better name, e.g. `space-in-array-notation`) and discussing the options it will support (e.g. ignoring nested objects, arrays, etc) as per the original `space-in-brackets` rule.

@nzakas once I've got feedback and direction from you on how you'd like this done, I'm happy to work on a PR implementing it.

### PR DESCRIPTION
**Pull request by xjamundx**
- Took the existing `space-in-brackets` rule and ripped out the `object` stuff.
- Added destructuring support

### CODE REVIEW

File: tests/lib/rules/array-bracket-spacing.js [hunk 1]
```diff
@@ -0,0 +1,764 @@
...
+
+        { code: "var foo = [{'bar':'baz'}, 1, {'bar': 'baz'}];", options: ["never"] },
+        { code: "var foo = [{'bar': 'baz'}];", options: ["never"] },
+        { code: "var foo = [{\n'bar': 'baz', \n'qux': [{'bar': 'baz'}], \n'quxx': 1 \n}]", options: ["never"] },
+        { code: "var foo = [1, {'bar': 'baz'}, 5];", options: ["never"] },
+        { code: "var foo = [{'bar': 'baz'}, 1,  5];", options: ["never"] },
+        { code: "var foo = [1, 5, {'bar': 'baz'}];", options: ["never"] },
+        { code: "var obj = {'foo': [1, 2]}", options: ["never"] },
+
+        // propertyName: false
```
Comment (xjamundx) [inline tests/lib/rules/array-bracket-spacing.js]: A bunch of these tests can be removed

File: docs/rules/array-bracket-spacing.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
+# Disallow or enforce spaces inside of brackets. (array-bracket-spacing)
+
+While formatting preferences are very personal, a number of style guides require or disallow spaces between array brackets:
+
+```js
+var obj = { foo: 'bar' };
```
Comment (xjamundx) [inline docs/rules/array-bracket-spacing.md]: I can remove some of these examples

**Comment by xjamundx** [comment]
I'll add some destructuring here as well...

File: docs/rules/array-bracket-spacing.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  "arraysInArrays": true
+}]
+```
+
+The following exceptions are available:
+
+* `singleValue` sets the spacing of a single value inside of square brackets of an array.
+* `objectsInArrays` sets the spacings between the curly braces and square brackets of object literals that are the first or last element in an array.
+* `arraysInArrays` sets the spacing between the square brackets of array literals that are the first or last element in an array.
+* `arraysInObjects` sets the spacing between the square bracket and the curly brace of an array literal that is the last element in an object.
```
Comment (xjamundx) [inline docs/rules/array-bracket-spacing.md]: needs to be removed

File: tests/lib/rules/array-bracket-spacing.js [hunk 4]
```diff
@@ -0,0 +1,764 @@
...
+
+var eslint = require("../../../lib/eslint"),
+    ESLintTester = require("eslint-tester");
+
+//------------------------------------------------------------------------------
+// Tests
+//------------------------------------------------------------------------------
+
+var eslintTester = new ESLintTester(eslint);
+eslintTester.addRuleTest("lib/rules/space-in-brackets", {
```
Comment (xjamundx) [inline tests/lib/rules/array-bracket-spacing.js]: every...single..time

Comment (xjamundx) [inline tests/lib/rules/array-bracket-spacing.js]: kind of wish i could jus pass in `__filename` and have the test work it out :D

**Comment by xjamundx** [comment]
Test fixes ASAP

**Comment by xjamundx** [comment]
@ilyavolodin was anything missing here?

**Comment by ilyavolodin** [comment]
LGTM

**Comment by nzakas** [comment]
Ack! This means I can't do a patch release. Let's be careful merging in new features the week after minor releases, as we usually need one or two patch releases to fix bugs.

**Comment by ilyavolodin** [comment]
Damn, I totally forgot about that, sorry!:-(

**Comment by mathieumg** [comment]
If that becomes a problem perhaps another branching model would be needed?

**Pull request merged**